examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2893 @@
+
+    template <typename T1, std::size_t k1, typename T2, typename T3>
+    DEAL_II_ALWAYS_INLINE inline void
+    scatter(std::array<T1, k1> &U, const T2 &result, const T3 i)
+    {
+      for (unsigned int j = 0; j < k1; ++j)
+        U[j].local_element(i) = result[j];
+    }
+  } // namespace
+
+  // We are now in a position to assemble all matrices stored in
+  // <code>OfflineData</code>: the lumped mass entries $m_i$, the
+  // vector-valued matrices $\mathbf{c}_{ij}$ and $\mathbf{n}_{ij} =
+  // \frac{\mathbf{c}_{ij}}{|\mathbf{c}_{ij}|}$, and the boundary normals
+  // $\boldsymbol{\nu}_i$.
+  //
+  // In order to exploit thread parallelization we use WorkStream approach
+  // detailed in the @ref threads Parallel computing with multiple processors
+  // accessing shared memory. As customary this requires
+  // definition of
+  //  - Scratch data (i.e. input info required to carry out computations): in
+  //    this case it is <code>scratch_data</code>.
+  //  - The worker: in the case it is <code>local_assemble_system()</code> that
+  //    actually computes the local (i.e. current cell) contributions from the
+  //    scratch data.
+  //  - A copy data: a struct that contains all the local assembly
+  //    contributions, in this case <code>CopyData<dim>()</code>.
+  //  - A copy data routine: in this case it is
+  //    <code>copy_local_to_global()</code> in charge of actually coping these
+  //    local contributions into the global objects (matrices and/or vectors)
+  //
+  // Most the following lines are spent in the definition of the worker
+  // <code>local_assemble_system()</code> and the copy data routine
+  // <code>copy_local_to_global()</code>. There is not much to say about the
+  // WorkStream framework since the vast majority of ideas are reasonably
+  // well-documented in step-9, step-13 and step-32 among others.
+  //
+  // Finally, assuming that $\mathbf{x}_i$ is a support point at the boundary,
+  // the (nodal) normals are defined as:
+  //
+  // @f{align*}
+  // \widehat{\boldsymbol{\nu}}_i :=
+  // \frac{\boldsymbol{\nu}_i}{|\boldsymbol{\nu}_i|} \ \text{where}
+  // \boldsymbol{\nu}_i := \sum_{T \subset \text{supp}(\phi_i)}
+  // \sum_{F \subset \partial T \cap \partial \Omega}
+  // \sum_{\mathbf{x}_{q,F}} \nu(\mathbf{x}_{q,F})
+  // \phi_i(\mathbf{x}_{q,F})
+  // @f}
+  //
+  // here $T$ denotes elements,
+  // $\text{supp}(\phi_i)$ the support of the shape function $\phi_i$,
+  // $F$ are faces of the element $T$, and $\mathbf{x}_{q,F}$
+  // are quadrature points on such face. Note that this formula for
+  // $\widehat{\boldsymbol{\nu}}_i$ is nothing else than some form of
+  // weighted averaging. Other more sophisticated definitions for $\nu_i$
+  // are possible but none of them have much influence in theory or practice.
+
+  template <int dim>
+  void OfflineData<dim>::assemble()
+  {
+    lumped_mass_matrix = 0.;
+    norm_matrix        = 0.;
```

Comment:
Is this matrix storing norms, or normals? If the latter, maybe call it `normal_matrix`?

Reply:
It is storing norms.